File: python/leetcodepy/copy_list_with_random_pointer.py
"""
A linked list is given such that each node contains an additional random pointer which could point to any node in the list or null.

Return a deep copy of the list.
"""
from collections import defaultdict

from .utils import RandomListNode


class Solution1(object):
    def copyRandomList(self, head):
        """
        :type head: RandomListNode
        :rtype: RandomListNode
        """
        def_dict = defaultdict(lambda: RandomListNode(0))
        def_dict[None] = None
        p = head
        while p:
            def_dict[p].label = p.label
            def_dict[p].next = def_dict[p.next]
            def_dict[p].random = def_dict[p.random]
            p = p.next
        return def_dict[head]

File: cpp/src/CopyListWithRandomPointer.cpp
#include "CopyListWithRandomPointer.h"

using namespace lcpp;

Node *Solution138_1::copyRandomList(Node *head) {
  if (head == nullptr)
    return nullptr;
  Node *P, *Copy;
  for (P = head; P != nullptr; P = Copy->next) {
    Copy = new Node(P->val);
    Copy->next = P->next;
    P->next = Copy;
  }
  for (P = head; P != nullptr; P = Copy->next) {
    Copy = P->next;
    if (P->random != nullptr)
      Copy->random = P->random->next;
  }
  auto CopyHead = head->next;
  for (P = head; P != nullptr;) {
    Copy = P->next;
    P->next = Copy->next;
    P = P->next;
    if (P != nullptr)
      Copy->next = P->next;
  }
  return CopyHead;
}
